### PR TITLE
ego coordinate system

### PR DESCRIPTION
**Pull request by YuechengLiu**

### CODE REVIEW

File: smarts-imitation/smarts_imitation/utils/adapter.py [hunk 1]
```diff
@@ -2,28 +2,39 @@
...
-def get_observation_adapter(neighbor_mode="LANE"):
-    closest_neighbor_num = 6
-    img_resolution = 40
-    observe_lane_num = 3
+def get_subscribed_features(neighbor_mode, closest_neighbor_num):
     subscribed_features = dict(
-        ego_pos=(2,),
-        heading=(1,),
-        speed=(1,),
+        ego_dynamics=(1,),
```
Comment (zbzhu99) [inline smarts-imitation/smarts_imitation/utils/adapter.py]: If the ego position is not included in the current observation, the `DistanceTravelled` metric can not be computed in eval_util.py.

File: smarts-imitation/smarts_imitation/envs/core.py [hunk 2]
```diff
@@ -54,7 +54,7 @@ def __init__(
         self.observation_space = gym.spaces.Box(
             low=-np.inf,
             high=np.inf,
-            shape=(28 * self.obs_stacked_size,),
+            shape=(12 * self.obs_stack_size,),
```
Comment (zbzhu99) [inline smarts-imitation/smarts_imitation/envs/core.py]: I suggest change the magic number here as an attribute from self.agent_spec.observation_adapter.

Now observation_adapter is a function, and we can change it as a class with `__call__` overrided with an additional attribute as obs_dim.

File: smarts-imitation/smarts_imitation/utils/adapter.py [hunk 3]
```diff
@@ -2,28 +2,39 @@
...
-def get_observation_adapter(neighbor_mode="LANE"):
-    closest_neighbor_num = 6
+def get_subscribed_features(neighbor_mode, closest_neighbor_num):
+    if neighbor_mode == "LANE":
+        subscribed_features = dict(
+            ego_pos=(2,),
+            heading=(1,),
+            speed=(1,),
+            neighbor_with_lanes=(closest_neighbor_num, 4),  # dist, speed, ttc
+        )
```
Comment (zbzhu99) [inline smarts-imitation/smarts_imitation/utils/adapter.py]: I have to keep this kind of 4-dimensional ego observation here to make sure some of the previous experiment results can be reproduced.

**Review by zbzhu99 — CHANGES_REQUESTED** [review]

File: ILSwiss/run_scripts/gen_smarts_demos.py [hunk 4]
```diff
@@ -229,16 +232,16 @@ def experiment(specs):
 
     with open(
         Path(save_path).joinpath(
-            "smarts_{}_stack-{}_{}.pkl".format(
+            "smarts_{}_stack-{}.pkl".format(
```
Comment (zbzhu99) [inline ILSwiss/run_scripts/gen_smarts_demos.py]: I do not quite understand how do we distinguish demonstration data with different feature_list?

**Comment by zbzhu99** [comment]
Also, plz format the code using `black` to make CI test pass :dizzy:, thanks~

(For more info about black, you can refer to https://black.readthedocs.io/en/stable/.)

**Comment by YuechengLiu** [comment]
> Also, plz format the code using `black` to make CI test pass 💫, thanks~
> 
> (For more info about black, you can refer to https://black.readthedocs.io/en/stable/.)

**Review by YuechengLiu — COMMENTED** [review]
great

**Review by zbzhu99 — APPROVED** [review]

**Pull request merged**